subsys/net/lib/http/http_server_http1.c

```diff
@@ -313,12 +313,86 @@ static int on_header_field(struct http_parser *parser, const char *at,
 	struct http_client_ctx *ctx = CONTAINER_OF(parser,
 						   struct http_client_ctx,
 						   parser);
+	size_t offset = strlen(ctx->header_buffer);
+
+	if (offset + length > sizeof(ctx->header_buffer) - 1U) {
+		LOG_DBG("Header %s too long (by %zu bytes)", "field",
+			offset + length - sizeof(ctx->header_buffer) - 1U);
+		ctx->header_buffer[0] = '\0';
+	} else {
+		memcpy(ctx->header_buffer + offset, at, length);
+		offset += length;
+		ctx->header_buffer[offset] = '\0';
+
+		if (parser->state == s_header_value_discard_ws) {
+			/* This means that the header field is fully parsed,
+			 * and we can use it directly.
+			 */
+			if (strncasecmp(ctx->header_buffer, "Upgrade",
+					sizeof("Upgrade") - 1) == 0) {
+				ctx->has_upgrade_header = true;
+			} else if (strncasecmp(ctx->header_buffer,
+					       "Sec-WebSocket-Key",
+					       sizeof("Sec-WebSocket-Key") - 1) == 0) {
+				ctx->websocket_sec_key_next = true;
+			}
+
+			ctx->header_buffer[0] = '\0';
+		}
+	}
 
 	ctx->parser_state = HTTP1_RECEIVING_HEADER_STATE;
 
 	return 0;
 }
 
+static int on_header_value(struct http_parser *parser,
+			   const char *at, size_t length)
+{
+	struct http_client_ctx *ctx = CONTAINER_OF(parser,
+						   struct http_client_ctx,
+						   parser);
+	size_t offset = strlen(ctx->header_buffer);
+
+	if (offset + length > sizeof(ctx->header_buffer) - 1U) {
+		LOG_DBG("Header %s too long (by %zu bytes)", "value",
+			offset + length - sizeof(ctx->header_buffer) - 1U);
+		ctx->header_buffer[0] = '\0';
+	} else {
+		memcpy(ctx->header_buffer + offset, at, length);
+		offset += length;
+		ctx->header_buffer[offset] = '\0';
+
+		if (parser->state == s_header_almost_done) {
+			if (ctx->has_upgrade_header) {
+				if (strncasecmp(ctx->header_buffer, "h2c",
+						sizeof("h2c") - 1) == 0) {
+					ctx->http2_upgrade = true;
+				} else if (strncasecmp(ctx->header_buffer,
+						       "websocket",
+						       sizeof("websocket") - 1) == 0) {
+					ctx->websocket_upgrade = true;
+				}
+
+				ctx->has_upgrade_header = false;
+			}
+
+			if (ctx->websocket_sec_key_next) {
+#if defined(CONFIG_WEBSOCKET)
+				strncpy(ctx->ws_sec_key, ctx->header_buffer,
+					MIN(sizeof(ctx->ws_sec_key),
+					    strlen(ctx->header_buffer)));
```

Comment:
The `MIN` macro would expand the `strlen` twice, can't use `offset` instead?

```suggestion
				strncpy(ctx->ws_sec_key, ctx->header_buffer,
					MIN(sizeof(ctx->ws_sec_key), offset));
```

Reply:
Good point, we could use offset here.